stablehlo/transforms/optimization/StablehloTargetIndependentOptimization.cpp

```diff
@@ -0,0 +1,66 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License, Version 2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// Implements optional canonicalization patterns for StableHLO ops.
+
+#include <cassert>
+#include <cstdint>
+#include <utility>
+
+#include "mlir/IR/BuiltinTypeInterfaces.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Rewrite/FrozenRewritePatternSet.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "stablehlo/transforms/optimization/Passes.h"
+
+namespace mlir {
+namespace stablehlo {
+
+#define GEN_PASS_DEF_STABLEHLOTARGETINDEPENDENTOPTIMIZATIONPASS
+#include "stablehlo/transforms/optimization/Passes.h.inc"
+
+// This is an upper limit on how many elements can be folded by an op folder.
+// This limit doesn't apply to some special cases like adding a zero,
+// multiplying by one, doing many operations with splats.
+constexpr int64_t kFoldOpEltLimit = 65536;
+
+struct StablehloTargetIndependentOptimizationPass
+    : public impl::StablehloTargetIndependentOptimizationPassBase<
+          StablehloTargetIndependentOptimizationPass> {
+  using StablehloTargetIndependentOptimizationPassBase::
+      StablehloTargetIndependentOptimizationPassBase;
+
+  LogicalResult initialize(MLIRContext* context) override {
+    RewritePatternSet patterns_(context);
+    bool foldFloat = false;
+    populateStablehloCanonicalizationPatterns(context, &patterns_);
+    populateStablehloAggressiveFolderPatterns(&patterns_, context, foldFloat,
```

Comment:
Are we missing any optimization if we keep the same benefits as 1 for both canonicalization and folding? If nit then how about we keep the benefit the same unless we find a case where we want canonicalization to happen before.

Reply:
We probably always want folding to have higher precedence than canonicalization.

There exist simplification patterns that rely on folding being applied (I have come across MLIR patterns that relied on the precedence of folding being higher than simplification in LiteRT). 

Also much of canonicalization relies on things having constants as inputs that it can reason about it, so folding first enables more canonicalization.